homeassistant/components/xiaomi_miio/__init__.py

```diff
@@ -175,12 +175,23 @@ async def async_create_miio_device_and_coordinator(
 
     async def async_update_data():
         """Fetch data from the device using async_add_executor_job."""
-        try:
+
+        async def _async_fetch_data():
+            """Fetch data from the device."""
             async with async_timeout.timeout(10):
                 state = await hass.async_add_executor_job(device.status)
                 _LOGGER.debug("Got new state: %s", state)
                 return state
 
+        try:
+            return await _async_fetch_data()
+        except DeviceException as ex:
+            if ex.code != -9999:
+                raise UpdateFailed(ex) from ex
+            _LOGGER.info("Got exception while fetching the state: %s", ex)
+        # Try to fetch the data a second time after error code -9999
+        try:
+            return await _async_fetch_data()
```

Comment:
It possibly can retry infinitely.

I think we should add a `retry=True` flag to be sure we are fetching only twice.

Reply:
If the first attempt ends with an exception, we raise `UpdateFailed()` (so we end this function) or we go to the second attempt if error code is `-9999`. Under what circumstances can there be an infinite retrying?

Reply:
Depends on how many times the device sends -9999. I also think it should be limited.

Reply:
The device returns this error for less than 10 seconds. Integration polls the device every 60 seconds. So there cannot be infinite retrying.